8276618: Pad cacheline for Thread::_rcu_counter

Hamlin-Li opened the pull request:
Currently, Thread::_rcu_counter is not padded by cacheline, it should be beneficail to do so.

The initial spebjbb test shows about 10.5% improvement of critical, and 0.7% improvement of max in specjbb2015, specjbb arguments:
  GROUP_COUNT=4
  TI_JVM_COUNT=1
  JAVA_OPTS_BE="-server -XX:+UseG1GC -Xms32g -Xmx32g"
  MODE_ARGS="-ikv"

Hamlin-Li commented:
/cc hotspot-gc

dholmes-ora commented on src/hotspot/share/runtime/thread.hpp:
```diff
@@ -250,10 +251,10 @@ class Thread: public ThreadShadow {
 
   // Support for GlobalCounter
  private:
```
pre-existing nit: this private is not needed; nor is the public at line 260.

dholmes-ora reviewed (commented):
Hi Hamlin,

This seems reasonable to me, however whenever we add padding to optimise the placement of one field, I always wonder if that same padding has de-optimised the placement of other fields? I think we need to see a broader run of benchmarks here and across more than just x86_64.

I will see if I can assist on the benchmark front.

Thanks,
David

Hamlin-Li commented:
Thanks a lot David, it will be very helpful.
BTW, I will modify as you suggested later together with other's comments.

tschatzl reviewed (commented):
I'll push it through our perf testing.

Hamlin-Li commented:
Thanks a lot Thomas. :)

tschatzl commented:
Hi,

  we tried to reproduce your numbers internally, but failed to do so. Differences seem to be within noise.
  
  We tried with specjbb2015 multi-jvm on a fairly large machine (152 threads; it was what has been "on hand") and multiple runs of our internal benchmarks and specjbb2015 composite runs on various (not-so-large but still fairly big sized machines).

Could you post or send more details about your configuration?

The other concern that has been brought up internally has been that this increases the size of Thread by ~40% from 624 to 872 bytes; do you think there a way to save some memory by reorganizing the fields so that the counter is on a separate cache line "naturally"?

Thanks,
  Thomas

Hamlin-Li commented:
Thanks Thomas for the feedback.

I have updated the summary of this PR with more configuration and environment info, and I also updated the 2nd round of run. Although the improvement is not that much as the previous round (10%), but still got about 3~4% improvement, and seems the data is more stable than the 1st round of run.
(JBS is not available currently, will update JBS too later)

Hamlin-Li commented:
> The other concern that has been brought up internally has been that this increases the size of Thread by ~40% from 624 to 872 bytes; do you think there a way to save some memory by reorganizing the fields so that the counter is on a separate cache line "naturally"?

Sure, if current change is proven to bring some performance benefit, let me do some more research in this direction.

Hamlin-Li commented:
need further investigation.
